app.py

```diff
@@ -1,26 +1,402 @@
+    """
+    Only allow given sentiments
+    """
+    sentiments = ["Positive", "Negative", "Uncertain"]
+    try:
+        val = str(val).title()
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} could not be parsed to a string")
+
+    if val not in sentiments:
+        raise argparse.ArgumentTypeError(f"{val} is not a valid sentiment. Possible values: {', '.join(sentiments)}")
+    return val
+
+
+def _restricted_plots(val: str):
+    """
+    Only allow given plots
+    """
+    plots = ["bar", "line", "pie"]
+    try:
+        val = str(val).lower()
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} could not be parsed to a string")
+
+    if val not in plots:
+        raise argparse.ArgumentTypeError(f"{val} is not a valid plot type. Possible values: {', '.join(plots)}")
+    return val
+
+
+def _restricted_hashtags(val: str):
+    """
+    Only allow hashtags that follow our standards.
+    Typically we remove #-symbols even if a user added them.
+    Used for hashtags arg
+    """
+    try:
+        val = str(val).lower()
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} could not be parsed to a string")
+
+    val = re.sub(_REGEX_CHAR_MATCHER_HASHTAGS, "", val)
+    return val
+
+
+def _restricted_search_hashtags(val: str):
+    """
+    Only allow hashtags that follow our standards.
+    Typically we add a # even if user forgot them.
+
+    Important to use .lower to match program output.
+
+
+    Used for search_hashtags arg
+    """
+    try:
+        val = str(val).lower()
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} could not be parsed to a string")
+
+    if not val.startswith('#'):
+        return '#' + val
+    return val
+
+
+def _restricted_search_mentions(val: str):
+    """
+    Only allow mentions that follow our standards.
+    Typically we add @-symbol even if an user forgot them.
+
+    important to keep structure to match program output.
+
+    Used for search_mentions arg
+    """
+    try:
+        val = str(val)
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} could not be parsed to a string")
+
+    if not val.startswith('@'):
+        return '@' + val
+    return val
+
+
+#########CUSTOM TYPES#########
+
+#########HELPER METHODS#########
+
+class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter, argparse.RawTextHelpFormatter):
+    """
+    Allows us to use two formatters for argparse instead of only the default 1 (hackish)
+    https://stackoverflow.com/a/18462760
+    """
+
+    def _split_lines(self, text, width):
+        # https://stackoverflow.com/a/29498894
+        # modify super to add newline between arguments
+        lines = super()._split_lines(text, width)
+        if text.endswith('\n'):  # custom newline
+            lines += ['']
+        else:
+            lines += ['']  # also add newline between different arguments
+        return lines
```

Comment:
What does this mean? 
What is a "formatter" in this context?

Reply:
The argparse constructor allows us to pass the following `formatter_class=argparse.HelpFormatter`.

The constructor only allows us to pass *one* formatter, despite several of the provided classes being helpful. This has been discussed and suggested to be changed several times, but argparse isn't really being updated. (probably why there are a lot of libraries that replaces/enhance argparse..)

As mentioned in this docstring, this simply allows us to use two formatters at the same time. It is very hackish.

On top of that, we modify the parser to split lines on every `\n` - something it doesn't do default (it is actually very hard to otherwise do a newline)

For descriptions of each formatter, see here:  
https://docs.python.org/3.8/library/argparse.html#formatter-class

Reply:
Alright. Thanks :) 
But what _is_ an argparse formatter? 
Is it "_just_" for formatting the help writeouts? 
![image](https://user-images.githubusercontent.com/11192971/82306351-3ed26300-99bf-11ea-9d77-42ff678cd797.png)

Reply:
Yeah, you could say so. :)

The main issue is that one of the classes formats the *help* part for the program itself, while the others format the *help* part of the arguments.

We want both, so it looks readable.  
Compared to the output I wrote in the above comment, without formatters it would look *something like this*:

```bash
$ python app.py -h
usage: TweetScraper9000
A program that scrapes Twitter for hashtags and performs a sentiment analysis on the results. Presents results in a chosen chart format.

positional arguments:
  hashtags        The hashtags to scrape. EXAMPLE: 'trump biden' -REQUIRED-

optional arguments:
  -h, --help            show this help message and exit
  -p PLOT_TYPE, --plot PLOT_TYPE
                        Plot chart type, choose one. VALUES=[bar, line, pie] (default: pie)
  -l, --local           Pass to attempt scraping from local files. (default: False)
  -s, --save            Pass to save plots locally (if omitted will show plots instead) (default: False)
```
which might look ok-readable here, but definitely is not in a terminal!